runtime/lua/vim/lsp/_watchfiles.lua

```diff
@@ -134,57 +134,53 @@ function M.register(reg, ctx)
   if not has_capability or not client.workspace_folders then
     return
   end
-  local watch_regs = {} --- @type table<string,{pattern:lpeg,kind:integer}>
-  for _, w in ipairs(reg.registerOptions.watchers) do
-    local relative_pattern = false
-    local glob_patterns = {} --- @type {baseUri:string, pattern: string}[]
-    if type(w.globPattern) == 'string' then
+  local register_options = reg.registerOptions --[[@as lsp.DidChangeWatchedFilesRegistrationOptions]]
+  ---@type table<string, {pattern: vim.lpeg.Pattern, kind: lsp.WatchKind}[]> by base_dir
+  local watch_regs = vim.defaulttable()
+  for _, w in ipairs(register_options.watchers) do
+    local kind = w.kind
+      or (protocol.WatchKind.Create + protocol.WatchKind.Change + protocol.WatchKind.Delete)
+    local glob_pattern = w.globPattern
+
+    if type(glob_pattern) == 'string' then
+      local pattern = parse(glob_pattern)
+      if not pattern then
+        error('Cannot parse pattern: ' .. glob_pattern)
+      end
       for _, folder in ipairs(client.workspace_folders) do
-        table.insert(glob_patterns, { baseUri = folder.uri, pattern = w.globPattern })
+        local base_dir = vim.uri_to_fname(folder.uri)
+        table.insert(watch_regs[base_dir], { pattern = pattern, kind = kind })
       end
     else
-      relative_pattern = true
-      table.insert(glob_patterns, w.globPattern)
-    end
-    for _, glob_pattern in ipairs(glob_patterns) do
-      local base_dir = nil ---@type string?
-      if type(glob_pattern.baseUri) == 'string' then
-        base_dir = glob_pattern.baseUri
-      elseif type(glob_pattern.baseUri) == 'table' then
-        base_dir = glob_pattern.baseUri.uri
-      end
-      assert(base_dir, "couldn't identify root of watch")
-      base_dir = vim.uri_to_fname(base_dir)
-
-      ---@type integer
-      local kind = w.kind
-        or protocol.WatchKind.Create + protocol.WatchKind.Change + protocol.WatchKind.Delete
-
+      local base_uri = glob_pattern.baseUri
+      local uri = type(base_uri) == 'string' and base_uri or base_uri.uri
+      local base_dir = vim.uri_to_fname(uri)
       local pattern = parse(glob_pattern.pattern)
-      assert(pattern, 'invalid pattern: ' .. glob_pattern.pattern)
-      if relative_pattern then
-        pattern = lpeg.P(base_dir .. '/') * pattern
+      if not pattern then
+        error('Cannot parse pattern: ' .. glob_pattern.pattern)
       end
-
-      watch_regs[base_dir] = watch_regs[base_dir] or {}
-      table.insert(watch_regs[base_dir], {
-        pattern = pattern,
-        kind = kind,
-      })
+      pattern = lpeg.P(base_dir .. '/') * pattern
```

Comment:
Are we sure that `base_dir` won't have a trailing slash?

Reply:
Not sure,  it's the same logic as before